bob/src/cluster/quorum.rs

```diff
@@ -351,25 +352,194 @@ impl Cluster for Quorum {
+
+        let mut result = vec![false; len];
+
+        let (local, mut primary, mut secondary) = group_by_nodes(keys, &self.mapper);
+
+        if let Some(local) = local {
+            if !local.is_empty() {
+                let local_keys = local.collect(keys);
+                let local_exist = exist_on_local_node(&self.backend, &local_keys).await?;
+                local.update_existence(&mut result, &local_exist);
+            }
+        }
+
+        collect_remote_exists(&mut result, keys, &mut primary).await?;
+        collect_remote_exists(&mut result, keys, &mut secondary).await?;
+
+        let alien_index_map = IndexMap::where_not_exists(&result);
+
+        if !alien_index_map.is_empty() {
+            let local_alien_result =
+                exist_on_local_alien(&self.backend, &alien_index_map.collect(keys)).await?;
+            alien_index_map.update_existence(&mut result, &local_alien_result);
+        }
+
+        if !alien_index_map.is_empty() {
+            let all_remote_nodes: Vec<_> = self
+                .mapper
+                .nodes()
+                .iter()
+                .filter(|n| n.name() != self.mapper.local_node_name())
+                .cloned()
+                .collect();
+            let remote_nodes_aliens_exist =
+                exist_on_remote_aliens(&all_remote_nodes, &alien_index_map.collect(keys)).await;
+            for remote_alien_result in remote_nodes_aliens_exist {
+                match remote_alien_result {
+                    Ok(remote_alien_result) => {
+                        alien_index_map.update_existence(&mut result, remote_alien_result.inner());
+                    }
+                    Err(e) => {
+                        debug!("Failed to check existence in aliens: {:?}", e);
+                    }
                 }
             }
         }
-        Ok(exist)
+
+        Ok(result)
     }
 
     async fn delete(&self, key: BobKey, meta: &BobMeta) -> Result<(), Error> {
         self.delete_on_nodes(key, meta).await
     }
 }
+
+async fn collect_remote_exists(
+    result: &mut [bool],
+    keys: &[BobKey],
+    indexes_by_node: &mut HashMap<Node, IndexMap>,
+) -> Result<(), Error> {
+    if !indexes_by_node.is_empty() {
+        let mut keys_by_node = HashMap::new();
+        for (node, node_map) in indexes_by_node.iter_mut() {
+            node_map.retain_not_existed(&result);
+            if !node_map.is_empty() {
+                keys_by_node.insert(node.clone(), node_map.collect(keys));
+            }
+        }
+
+        if !keys_by_node.is_empty() {
+            let nodes: Vec<_> = keys_by_node.keys().cloned().collect();
+            let remote_results = exist_on_remote_nodes(&nodes, keys_by_node).await;
+            for (remote_result, node) in remote_results.into_iter().zip(nodes) {
+                match remote_result {
+                    Ok(remote_result) => {
+                        indexes_by_node
+                            .get(&node)
```

Comment:
The node should be presented. This should be enforced by algorithm, thus it is better to use `expect()` to access the `IndexMap` for node

Reply:
Additionally, add `debug_assert!(node.name() == remote_result.node_name())`